sstables/sstable_set.cc

```diff
@@ -61,94 +61,422 @@ std::ostream& operator<<(std::ostream& os, const sstables::sstable_run& run) {
+    return _erased;
+}
+
+const std::unordered_set<sstable_set_version*>& sstable_set_version::get_next_versions() const {
+    return _next;
+}
+
+unsigned sstable_set_version::get_reference_count() const {
+    return _reference_count;
+}
+
+schema_ptr sstable_set_version::get_schema() const {
+    return _schema;
+}
+
+std::vector<shared_sstable> sstable_set_version::select(const dht::partition_range& range) const {
+    return boost::copy_range<std::vector<shared_sstable>>(_versioned_data->select(range)
+         | boost::adaptors::filtered([this] (shared_sstable sst) { return this->contains(sst); }));
+}
+
+// Return all runs which contain any of the input sstables.
+std::vector<sstable_run> sstable_set_version::select_sstable_runs(const std::vector<shared_sstable>& sstables) const {
+    auto run_ids = boost::copy_range<std::unordered_set<utils::UUID>>(sstables | boost::adaptors::transformed(std::mem_fn(&sstable::run_identifier)));
+    return boost::copy_range<std::vector<sstable_run>>(run_ids | boost::adaptors::transformed([this] (utils::UUID run_id) {
+        return sstable_run(boost::copy_range<std::unordered_set<shared_sstable>>(_versioned_data->select_by_run_id(run_id)
+                         | boost::adaptors::filtered([this] (shared_sstable sst) { return this->contains(sst); })));
+    }));
+}
+
+const std::set<shared_sstable>& sstable_set_version::all() const {
+    return _versioned_data->all();
+}
+
+bool sstable_set_version::contains(shared_sstable sst) const { 
+    return _added.contains(sst) || (!_erased.contains(sst) && _prev && _prev->contains(sst));
+}
+
+size_t sstable_set_version::size() const {
+    return _added.size() - _erased.size() + (_prev ? _prev->size() : 0);
+}
+
+void sstable_set_version::insert(shared_sstable sst) {
+    if (this->contains(sst)) {
+        return;
+    }
+    if (_erased.contains(sst)) {
+        _erased.erase(sst);
+    } else {
+        _versioned_data->insert(sst);
+        try {
+            _added.insert(sst);
+        } catch (...) {
+            _versioned_data->remove(sst);
+            throw;
+        }
+    }
+}
+
+void sstable_set_version::erase(shared_sstable sst) {
+    if (!this->contains(sst)) {
+        return;
+    }
+    if (_added.contains(sst)) {
+        _added.erase(sst);
+        _versioned_data->remove(sst);
+    } else {
+        _erased.insert(sst);
+        if (_prev->get_reference_count() == _prev->get_next_versions().size()) {
+            for (auto ver : _prev->get_next_versions()) {
+                if (!ver->get_sstables_erased().contains(sst)) {
+                    return;
+                }
+            }
+            // we may be able to release the sstable
+            for (auto ver : _prev->get_next_versions()) {
+                ver->unerase(sst);
+            }
+            _prev->erase(sst);
+        }
+    }
+}
+
+void sstable_set_version::unerase(shared_sstable sst) {
```

Comment:
Isn't this equivalent to `insert()`?

Reply:
Looks like it is the same in practice, I'll remove this one